### PR TITLE
add new env: Deepcoder

### PR DESCRIPTION
**Pull request by EarthRecovery**
feat: Add new env deepcoder

### CODE REVIEW

File: ragen/env/deepcoder/env.py [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+                "action_is_valid": False,
+                "success": False,
+                "error": "Empty action",
+            }
+            self.last_feedback = info["error"]
+            if self.current_prompt:
+                self.render_cache = f"{self.current_prompt}\n\nFeedback: {self.last_feedback}"
+            else:
+                self.render_cache = observation
+            return observation, reward, done, info
```
[the remark this answers is not in the document]
Comment (YuxiangLin-kaslana) [reply ragen/env/deepcoder/env.py:106]: This doesn’t make sense, because if an action is invalid, the environment should return a reward of 0 by default.

File: ragen/env/deepcoder/run.sh [hunk 2]
```diff
@@ -0,0 +1,28 @@
+#!/usr/bin/env bash
+set -e
+
+cd /workspace/RAGEN_temp
+export PYTHONPATH="$PWD:$PWD/verl"
```
[the remark this answers is not in the document]
Comment (YuxiangLin-kaslana) [reply ragen/env/deepcoder/run.sh]: fixed

File: ragen/env/deepcoder/utils.py [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+            if isinstance(raw_input, str):
+                parts = [p for p in raw_input.strip().splitlines() if p.strip()]
+                try:
+                    args = [eval(p) for p in parts]
+                except Exception:
+                    args = [raw_input]
+            elif isinstance(raw_input, (list, tuple)):
+                args = list(raw_input)
+            else:
+                args = [raw_input]
```
[the remark this answers is not in the document]
Comment (YuxiangLin-kaslana) [reply ragen/env/deepcoder/utils.py:147]: This code will not be run in the server so the secure issue will not be concerned

File: ragen/env/deepcoder/utils.py [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+                    "        print(json.dumps({\"ok\": True, \"result\": result}, ensure_ascii=False))",
+                    "    except Exception as e:",
+                    "        print(json.dumps({\"ok\": False, \"error\": repr(e)}))",
+                    "if __name__ == \"__main__\":",
+                    "    _run()",
+                    "",
+                ]
+            )
+
+            print(harness)
```
[the remark this answers is not in the document]
Comment (YuxiangLin-kaslana) [reply ragen/env/deepcoder/utils.py]: fixed

File: ragen/env/deepcoder/env.py [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+        try:
+            train_dataset, test_dataset = prepare_deepcoder_data()
+            dataset_bundle = {"train": train_dataset, "test": test_dataset}
+        except Exception:
+            dataset_bundle = None
+
+        if dataset_bundle is not None:
+            self.dataset = dataset_bundle
+        else:
+            self.dataset = None
```
[the remark this answers is not in the document]
Comment (YuxiangLin-kaslana) [reply ragen/env/deepcoder/env.py:44]: fixed

File: ragen/env/deepcoder/env.py [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+    def _sample_problem(self, seed: Optional[int] = None) -> Tuple[str, str]:
+        if self.dataset is None:
+            prompt = "Write a function add(a, b) that returns a + b."
+            solution = "def add(a, b):\n    return a + b"
+            return prompt, solution
+
+        split = self.config.split or next(iter(self.dataset.keys()))
+        dataset_split = self.dataset[split]
+        index = random.randint(0, len(dataset_split) - 1)
+        item = dataset_split[index]
```
[the remark this answers is not in the document]
Comment (YuxiangLin-kaslana) [reply ragen/env/deepcoder/env.py]: fixed

File: ragen/env/deepcoder/run.sh [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+python train.py --config-name _11_deepcoder \
+    model_path="Qwen/Qwen2.5-3B-Instruct" \
+    trainer.project_name=deepcoder_P1 \
+    trainer.experiment_name=deepcoder_100turns_test \
+    trainer.total_training_steps=100 \
+    ppo_mini_batch_size=4 \
+    micro_batch_size_per_gpu=1 \
+    es_manager.train.env_groups=32 es_manager.train.group_size=1 es_manager.train.env_configs.n_groups=[32] \
+    es_manager.val.env_groups=1 es_manager.val.group_size=1 es_manager.val.env_configs.n_groups=[1] \
+    system.CUDA_VISIBLE_DEVICES=\"0,1,2,3\" trainer.n_gpus_per_node=4 actor_rollout_ref.rollout.tensor_model_parallel_size=4 \
```
[the remark this answers is not in the document]
Comment (YuxiangLin-kaslana) [reply ragen/env/deepcoder/run.sh:18]: Users can configure this setting based on their device specifications.

File: ragen/env/deepcoder/env.py [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+
+    def _sample_problem(self, seed: Optional[int] = None) -> Tuple[str, str]:
+        if self.dataset is None:
+            prompt = "Write a function add(a, b) that returns a + b."
+            solution = "def add(a, b):\n    return a + b"
+            return prompt, solution
+
+        available_splits = list(self.dataset.keys())
+        split = self.config.split if self.config.split in self.dataset else available_splits[0]
+        dataset_split = self.dataset[split]
```
[the remark this answers is not in the document]
Comment (YuxiangLin-kaslana) [reply ragen/env/deepcoder/env.py]: fixed

File: ragen/env/deepcoder/env.py [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
```
[the remark this answers is not in the document]
Comment (YuxiangLin-kaslana) [reply ragen/env/deepcoder/env.py:54]: In our supported data-loading paths, self.dataset is guaranteed to have at least one split (train/test bundle or load_dataset failure before this point), so available_splits[0] is safe under current invariants
